### PR TITLE
Clean up window partition

### PR DESCRIPTION
**Pull request by martint**
The WindowPartition class got too complicated with the introduction of RANGE and GROUPS framing. This splits out the behavior of each framing scheme into separate classes.

### CODE REVIEW

File: core/trino-main/src/main/java/io/trino/operator/window/WindowPartition.java [hunk 1]
```diff
@@ -109,59 +77,58 @@ public WindowPartition(
...
+                            frame,
+                            partitionStart,
+                            partitionEnd,
+                            pagesIndex,
+                            peerGroupHashStrategy,
+                            0,
+                            0,
+                            peerGroupEnd - partitionStart - 1,
+                            0);
+                    break;
```
Comment (kasiafi) [inline core/trino-main/src/main/java/io/trino/operator/window/WindowPartition.java:125]: Do we need to pass the `0`s?

File: core/trino-main/src/main/java/io/trino/operator/window/RowsFraming.java [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+        return toIntExact(rowPosition + value);
+    }
+
+    private long getValue(int channel, int currentPosition)
+    {
+        checkState(!pagesIndex.isNull(channel, currentPosition), "Window frame offset must not be null");
+        long value = pagesIndex.getLong(channel, currentPosition);
+        checkState(value >= 0, "Window frame offset must not be negative");
+        return value;
+    }
```
Comment (kasiafi) [inline core/trino-main/src/main/java/io/trino/operator/window/RowsFraming.java:143]: Apparently, you didn't hit any test with this changed error message. We might be missing something.

**Review by kasiafi — COMMENTED** [review]
Thanks for working on this!

I was thinking about another possible optimization: gather window functions with matching framings and do the frame bound computations once for a group. After this cleanup it will be much easier to do.

File: core/trino-main/src/main/java/io/trino/operator/window/RowsFraming.java (hunk 2 above)
Comment (kasiafi) [reply core/trino-main/src/main/java/io/trino/operator/window/RowsFraming.java:143]: Ah, it's a checkstate... The actual check is done elsewhere. But a test is missing anyway.

**Comment by kasiafi** [comment]
In the constructors of `RowsFraming`, `GroupsFraming` and `RangeFraming`, you could add a check that the passed `FrameInfo` is consistent in frame type with the created Framing.

File: core/trino-main/src/main/java/io/trino/operator/window/GroupsFraming.java [hunk 3]
```diff
@@ -33,9 +35,9 @@
     private final Function<Integer, Integer> seekGroupStart;
     private final Function<Integer, Integer> seekGroupEnd;
 
-    // Recently computed frames for functions with frame type GROUPS.
-    // Along frame start and frame end, they also capture indexes of peer groups
-    // where frame bounds fall.
+    // Recently computed frame
+    // Along frame start and frame end, it alsos captures indexes of peer groups
+    // where frame bounds falls.
```
Comment (kasiafi) [inline core/trino-main/src/main/java/io/trino/operator/window/GroupsFraming.java:40]: typo in "also",
falls -> fall

**Review by kasiafi — COMMENTED** [review]
Looks good to me.

**Review by dain — APPROVED** [review]

**Pull request merged**